File: packages/kits/AriaMcpKit/Sources/AriaMCP/SurfacedRecallLedger.swift
// SurfacedRecallLedger.swift — per-session record of drawer ids surfaced to the
// AI client by `moot_memory_search`.
//
// The ledger is the ARIA boundary implementation of the trace-reward wiring
// (DESIGN_TRACE_REWARD_2026-06-12). When `moot_memory_search` surfaces a
// drawer, its id is recorded here with the wall-clock time of the recall.
// When a subsequent dereference verb (`moot_withdraw_memory`,
// `moot_update_memory`, `moot_confirm_memory`, `moot_move_memory`) acts on a
// drawer id that is present in the ledger, `noteUsage` in ToolDispatch calls
// `kit.markRecallUsed` so the dreaming daemon's reward sweep later assigns
// reward 1.0.
//
// Scope of "used" (option E from the design memo = B∪C):
//   - "surfaced" = the id appeared in a `moot_memory_search` result within
//     the current stdio / HTTP session.
//   - "used" = the same id appears in a later dereference call.
//   - Graph verbs (`moot_link_memories`, `moot_connection_map`) are NOT
//     dereference verbs in v1 (read-only blind spot; accepted per design).
//
// Session scope: one ledger per `ToolDispatcher` instance, i.e., one per
// stdio connection or HTTP session. Entries are never pruned during the
// session; memory footprint is bounded by session length × average result
// set size.
//
// Structural difference from Rust: Rust wraps a `HashMap` in a `Mutex`
// for interior mutability on `&self`. Swift uses an `actor` for the same
// guarantee — serial execution, no data races, Sendable conformance.
// Behavioral semantics are identical: `recordSurfaced` overwrites any
// existing entry for the same id (last surfaced_at wins), `entry(for:)`
// returns nil when the id was never surfaced.

import Foundation

/// One entry in the session ledger.
struct LedgerEntry: Sendable {
    /// The drawer id that was surfaced.
    let drawerID: String
    /// Wall-clock time when the drawer was surfaced by `moot_memory_search`.
    /// Stored as a `Date` so callers can pass it directly to GLK.
    let surfacedAt: Date
}

/// Session-scoped ledger of drawer ids that were returned to the AI client
/// by `moot_memory_search`. Thread-safe via Swift actor isolation.
///
/// Use `recordSurfaced(_:at:)` after a `moot_memory_search` call.
/// Use `entry(for:)` from a dereference verb to retrieve the entry that
/// enables reward-trace marking.
actor SurfacedRecallLedger: Sendable {

    // map: drawerID → LedgerEntry (last surfacedAt wins on duplicates)
    private var inner: [String: LedgerEntry] = [:]

    /// Create a new empty ledger.
    init() {}

    /// Record that `drawerIDs` were just surfaced by `moot_memory_search`.
    ///
    /// `at` is the wall-clock time of the recall. Any existing entry for the
    /// same id is overwritten with the latest `surfacedAt` so the retention
    /// window stays current.
    func recordSurfaced(_ drawerIDs: [String], at surfacedAt: Date) {
        for id in drawerIDs {
            inner[id] = LedgerEntry(drawerID: id, surfacedAt: surfacedAt)
        }
    }

    /// Look up the ledger entry for `drawerID`. Returns `nil` if the id was
    /// never surfaced in this session.
    func entry(for drawerID: String) -> LedgerEntry? {
        inner[drawerID]
    }

    /// Returns true if `drawerID` was surfaced in this session.
    func contains(_ drawerID: String) -> Bool {
        inner[drawerID] != nil
    }
}

File: packages/kits/AriaMcpKit/rust/src/surfaced_recall_ledger.rs
//! SurfacedRecallLedger — per-session record of drawer ids surfaced to the
//! AI client by `moot_memory_search`.
//!
//! The ledger is the ARIA boundary implementation of the trace-reward wiring
//! design (DESIGN_TRACE_REWARD_2026-06-12.md). When `moot_memory_search`
//! surfaces a drawer, its id is recorded here with the wall-clock time of the
//! recall. When a subsequent dereference verb (`moot_withdraw_memory`,
//! `moot_update_memory`, `moot_confirm_memory`, `moot_move_memory`) acts on
//! a drawer id that is present in the ledger, this module calls
//! `coordinator.mark_recall_used` on that id so the dreaming daemon's reward
//! sweep later assigns reward 1.0.
//!
//! Scope of "used" (option E from the design memo = B∪C):
//!   - "surfaced" = the id appeared in a `moot_memory_search` result within
//!     the current stdio session.
//!   - "used" = the same id appears in a later dereference call.
//!   - Graph verbs (`moot_link_memories`, `moot_connection_map`) are NOT
//!     dereference verbs in v1 (read-only blind spot; accepted per design).
//!
//! The ledger is session-scoped (one per Dispatcher instance, i.e. one per
//! stdio connection). Entries are never pruned during the session; the memory
//! footprint is bounded by the session length × average result set size.
//! Entries carry the surfaced_at timestamp so `mark_recall_used` can pass the
//! correct `now` bound to the retention-window query.
//!
//! Interior mutability: the ledger is wrapped in `Mutex` so the `Dispatcher`
//! (which takes `&self`) can mutate session state without requiring `&mut`.

use std::collections::HashMap;
use std::sync::Mutex;

/// One entry in the ledger: the wall-clock time (seconds since Unix epoch)
/// when the drawer was surfaced to the client.
#[derive(Debug, Clone)]
pub struct LedgerEntry {
    /// The drawer id that was surfaced.
    pub drawer_id: String,
    /// Wall-clock time (seconds since Unix epoch, i64) when the drawer was
    /// surfaced by `moot_memory_search`. Used as the `now` upper-bound when
    /// `mark_recall_used` is called (the trace rows' `recalledAt` should be
    /// close to but not after this value in a real session; we pass `now +1`
    /// to account for sub-second jitter).
    pub surfaced_at_secs: i64,
}

/// Session-scoped ledger of drawer ids that were returned to the AI client
/// by `moot_memory_search`. Thread-safe via internal `Mutex`.
///
/// Use `record_surfaced` after a `moot_memory_search` call.
/// Use `note_usage` from a dereference verb to trigger the reward mark.
#[derive(Debug, Default)]
pub struct SurfacedRecallLedger {
    // map: drawer_id → LedgerEntry (last surfaced_at wins on duplicates)
    inner: Mutex<HashMap<String, LedgerEntry>>,
}

impl SurfacedRecallLedger {
    /// Create a new empty ledger.
    pub fn new() -> Self {
        Self::default()
    }

    /// Record that `drawer_ids` were just surfaced by `moot_memory_search`.
    ///
    /// `surfaced_at_secs` is the wall-clock time of the recall (Unix epoch
    /// seconds). Any existing entry for the same id is overwritten with the
    /// latest surfaced_at so the retention window stays current.
    pub fn record_surfaced(&self, drawer_ids: &[String], surfaced_at_secs: i64) {
        if let Ok(mut map) = self.inner.lock() {
            for id in drawer_ids {
                map.insert(
                    id.clone(),
                    LedgerEntry {
                        drawer_id: id.clone(),
                        surfaced_at_secs,
                    },
                );
            }
        }
    }

    /// Look up the ledger entry for `drawer_id`. Returns `None` if the id
    /// was never surfaced in this session.
    pub fn get(&self, drawer_id: &str) -> Option<LedgerEntry> {
        self.inner
            .lock()
            .ok()
            .and_then(|map| map.get(drawer_id).cloned())
    }

    /// Returns true if `drawer_id` was surfaced in this session.
    pub fn contains(&self, drawer_id: &str) -> bool {
        self.inner
            .lock()
            .ok()
            .map(|map| map.contains_key(drawer_id))
            .unwrap_or(false)
    }
}
